watchers/common.go

```diff
@@ -16,32 +15,41 @@ const maxWaitPeriod = 30
 type WatchConsumer func(<-chan watch.Event) error
 
 // watchFunc manages object watching.
-type watchFunc func(context.Context, time.Duration) error
+type watchFunc func(context.Context) error
 
 // start wathing given resources.
-func start(ctx context.Context, name string, watchFunc watchFunc) {
+func start(ctx context.Context, name string, watchFunc watchFunc) <-chan error {
+	errChan := make(chan error)
+
 	go func() {
 		log := ctrl.Log.WithName(name + " watcher")
 
-		for waitPeriod := 0; ; waitPeriod = int(math.Min(float64((waitPeriod+1)*2), float64(maxWaitPeriod))) {
-			waitDuration := time.Second * time.Duration(waitPeriod)
+		for waitPeriod := 0; ; waitPeriod = (waitPeriod + 1) * 2 {
```

Comment:
If you are not using the 3 params of a for loop statement you don't need it, instead
```go
waitPeriod := 0
for {
    ...
    waitPeriod = (waitPeriod + 1) * 2
}
```
Makes it a lot easier to read

Reply:
Not agree, this doesn't looks more readable:
```
waitPeriod := 0
for {
			if waitPeriod != 0 {
				log.Info("wait before start", "period", waitPeriod)
				time.Sleep(time.Second * time.Duration(waitPeriod))
			}

			err := watchFunc(ctx)
			if err != nil {
				if waitPeriod >= maxWaitPeriod {
					log.Error(err, "time to give up watching")
					errChan <- err
					break
				}

				log.Error(err, "there was an error during watch")
				waitPeriod = (waitPeriod + 1) * 2
				continue
			}

			// Watch should be stop.
			log.Info("watch has ended")
			close(errChan)
			break
		}
```

Reply:
```go
waitPeriod := 0
for {
    if err := watchFunc(ctx); err == nil {
        log.Info("watch has ended")
        close(errChan) break
    }

    if watcherRecoveryPeriod >= maxWatcherRecoveryPeriod {
        log.Error(err, "watcher not able to recovered from failure")
        errChan <- err
        break
    }

    log.Error(err, "there was an error while watching...something") // add what this is watching
    waitPeriod = (waitPeriod + 1) * 2
    time.Sleep(time.Second * time.Duration(waitPeriod))
}
```

Reply:
This is what I meant, I also simplified the error handling to return earlier when possible and reducing the nesting code